StartsWith Linux perf improvements

adamsitnik opened the pull request:
This PR improves the performance of Culture-aware `string.StartsWith` on Linux.

Unfortunately, ICU does not expose a method that allows for optimal `StartsWith` using Collator API (there is a `StartsWith` that allows for comparing `UnicodeStrings` but without a possibility to specify culture).

I read some docs, articles and studied the code of ICU itself and came up with this proposal.

The longer the source text and the more unlucky case (miss) the bigger the gains.

It should help a lot with https://github.com/dotnet/corefx/issues/40674

```cs
public class Perf_StartsWith
{
    public IEnumerable<object[]> StartsWithArguments()
    {
        yield return new object[] { "a", "a" };
        yield return new object[] { "aaaaaaaaaaa", "a" };
        yield return new object[] { "a", "aaaaaaaaaaa" };
        yield return new object[] { new string('a', 512), "a" };
        yield return new object[] { new string('a', 512), "aaaaaaaaaaa" };
        yield return new object[] { new string('a', 512), new string('a', 512) };
        yield return new object[] { new string('a', 2048), "a" };
        yield return new object[] { new string('a', 2048), "aaaaaaaaaaa" };
        yield return new object[] { new string('a', 2048), new string('a', 2048) };
    }

    [GlobalSetup]
    public void Setup() => Thread.CurrentThread.CurrentCulture = CultureInfo.GetCultureInfo("fr-FR"); // isAsciiEqualityOrdinal=false

    [Benchmark]
    [ArgumentsSource(nameof(StartsWithArguments))]
    public bool StartsWith(string text, string prefix) => text.StartsWith(prefix);
}
```

|     Method |          Toolchain |                   text |                 prefix |         Mean | Ratio |
|----------- |------------------- |----------------------- |----------------------- |-------------:|------:|
| StartsWith |     /after/corerun |                      a |                      a |     2.930 ns |  1.04 |
| StartsWith |    /before/corerun |                      a |                      a |     2.807 ns |  1.00 |
|            |                    |                        |                        |              |       |
| StartsWith |     /after/corerun |                      a |            aaaaaaaaaaa |   325.073 ns |  0.05 |
| StartsWith |    /before/corerun |                      a |            aaaaaaaaaaa | 6,422.332 ns |  1.00 |
|            |                    |                        |                        |              |       |
| StartsWith |     /after/corerun |            aaaaaaaaaaa |                      a |   362.019 ns |  0.06 |
| StartsWith |    /before/corerun |            aaaaaaaaaaa |                      a | 5,589.542 ns |  1.00 |
|            |                    |                        |                        |              |       |
| StartsWith |     /after/corerun |  aaaaa(...)aaaaa [512] |                      a |   510.155 ns |  0.08 |
| StartsWith |    /before/corerun |  aaaaa(...)aaaaa [512] |                      a | 6,032.578 ns |  1.00 |
|            |                    |                        |                        |              |       |
| StartsWith |     /after/corerun |  aaaaa(...)aaaaa [512] |            aaaaaaaaaaa |   664.491 ns |  0.09 |
| StartsWith |    /before/corerun |  aaaaa(...)aaaaa [512] |            aaaaaaaaaaa | 7,143.208 ns |  1.00 |
|            |                    |                        |                        |              |       |
| StartsWith |     /after/corerun |  aaaaa(...)aaaaa [512] |  aaaaa(...)aaaaa [512] |     2.934 ns |  1.03 |
| StartsWith |    /before/corerun |  aaaaa(...)aaaaa [512] |  aaaaa(...)aaaaa [512] |     2.844 ns |  1.00 |
|            |                    |                        |                        |              |       |
| StartsWith |     /after/corerun | aaaaa(...)aaaaa [2048] |                      a |   551.200 ns |  0.08 |
| StartsWith |    /before/corerun | aaaaa(...)aaaaa [2048] |                      a | 6,754.139 ns |  1.00 |
|            |                    |                        |                        |              |       |
| StartsWith |     /after/corerun | aaaaa(...)aaaaa [2048] |            aaaaaaaaaaa |   760.843 ns |  0.11 |
| StartsWith |    /before/corerun | aaaaa(...)aaaaa [2048] |            aaaaaaaaaaa | 7,071.727 ns |  1.00 |
|            |                    |                        |                        |              |       |
| StartsWith |     /after/corerun | aaaaa(...)aaaaa [2048] | aaaaa(...)aaaaa [2048] |     2.787 ns |  0.96 |
| StartsWith |    /before/corerun | aaaaa(...)aaaaa [2048] | aaaaa(...)aaaaa [2048] |     2.876 ns |  1.00 |

GrabYourPitchforks commented:
I don't think the ordinal optimization is valid when performing a culture-aware comparison.

Example: consider a thread with current culture "hu-HU". The result of `"dz".StartsWith("d", StrinComparison.CurrentCulture)` should be _false_.

adamsitnik commented:
@GrabYourPitchforks thanks for pointing this out! I've created https://github.com/dotnet/corefx/pull/41016 to make sure we don't break it

tarekgh commented:
Thanks @adamsitnik for working on that, 

Can we hold a little on this change. It need to be reviewed carefully as I am not sure if there is specific cases need to be checked manually for accuracy. ligatures is one example of that. Also, we are not distinguishing between the error cases and the end of the string cases. I am not sure how this handled here, that is why need to look carefully to the change.

Side point, we are adding more complex code which will make touching this code or related code be more challenging. We still have the option try to optimize ICU itself if we are really seeing this is very critical. 

I am not really pushing back on the change, I am just trying to avoid going to issues that maybe not clear to us now.

tarekgh commented:
CC @eerhardt

tarekgh commented on src/corefx/System.Globalization.Native/pal_collation.c:
```diff
@@ -588,47 +592,153 @@ int32_t GlobalizationNative_IndexOfOrdinalIgnoreCase(
+        case UCOL_SECONDARY:
+            return UCOL_PRIMARYORDERMASK | UCOL_SECONDARYORDERMASK;
+        default:
+            return UCOL_PRIMARYORDERMASK | UCOL_SECONDARYORDERMASK | UCOL_TERTIARYORDERMASK;
```
Theoretically we shouldn't get strength different than the primary and the secondary. You may add assert in the default case.

adamsitnik replied:
I am not sure about this. When we call  `ucol_open(lpLocaleName, &err)` it returns a collator with strength == 2 (tertiary) and later we just clone it without setting the strength in explicit way

tarekgh replied:
The following code just print ```strength = UCOL_PRIMARY```. where did you see it returns tertiary

```C#
static void PrintDefaultCollator()
{
    UErrorCode err = U_ZERO_ERROR;
    UCollator* pCollator = ucol_open("en_US", &err);
    if (U_FAILURE(err))
    {
        printf("ucol_open('en_US', ...) failed with error %d\n", err);
        return;
    }

    UColAttributeValue attributeValue = ucol_getAttribute(pCollator, UCOL_STRENGTH, &err);
    if (!U_SUCCESS(err))
    {
        printf("ucol_getAttribute(...) failed with error %d\n", err);
        return;
    }

    switch (attributeValue)
    {
        case UCOL_PRIMARY: printf("strength = UCOL_PRIMARY\n"); break;
        case UCOL_SECONDARY: printf("strength = UCOL_SECONDARY\n"); break;
        case UCOL_TERTIARY: printf("strength = UCOL_PRIMARY\n"); break;
        default: printf("strength = %d\n", attributeValue);
    }
}
```

tarekgh replied:
by the way UCOL_PRIMARY = 2

tarekgh replied:
never mind. you are right.

tarekgh replied:
Maybe the assert should be on the UCOL_PRIMARY case then, because you'll always will have secondary and tertiary only. I was confused by the default strength. And you'll not need the check 

```
strength >= UCOL_SECONDARY inside SimpleStartsWith_Iterators.
```

tarekgh commented on src/corefx/System.Globalization.Native/pal_collation.c:
```diff
@@ -588,47 +592,153 @@ int32_t GlobalizationNative_IndexOfOrdinalIgnoreCase(
-        int32_t idx = USEARCH_DONE;
+        result = ucol_next(pIterator, pErrorCode);
+    }
+    while (result == UCOL_IGNORABLE); // we don't check errorCode because on error the result is set to UCOL_NULLORDER
```
>while (result == UCOL_IGNORABLE); // we don't check errorCode because on error the result is set to UCOL_NULLORDER [](start = 4, length = 114)

not checking the error code can cause a weird behavior if it happen. this method will return UCOL_NULLORDER which can make SimpleStartsWith_Iterators return true and that is wrong. may be you just OR all error codes and then check it one time at the end?

adamsitnik replied:
very good point, I am going to add a check for that  to the calling method

tarekgh commented on src/corefx/System.Globalization.Native/pal_collation.c:
```diff
@@ -588,47 +592,153 @@ int32_t GlobalizationNative_IndexOfOrdinalIgnoreCase(
-                else
-                {
-                    result = CanIgnoreAllCollationElements(pColl, lpSource, idx);
+                    return FALSE; // the next non-ignorable character in source text is a combining character
```
>return FALSE; // the next non-ignorable character in source text is a combining character [](start = 20, length = 89)

Do you have example of this case?

adamsitnik replied:
Yes, `CultureInfo.InvariantCulture.CompareInfo.IsPrefix("o\u0308", "o")` provided by @GrabYourPitchforks

tarekgh commented on src/corefx/System.Globalization.Native/pal_collation.c:
```diff
@@ -588,47 +592,153 @@ int32_t GlobalizationNative_IndexOfOrdinalIgnoreCase(
+    }
+}
+
+int32_t SimpleStartsWith(const UCollator* pCollator, UErrorCode* pErrorCode, const UChar* pPattern, int32_t patternLength, const UChar* pText, int32_t textLength)
```
>SimpleStartsWith [](start = 8, length = 16)

Did you test this logic with the Surrogate characters (well and malformed characters)? I am not expecting any problem but just want to ensure we are not missing any case.

adamsitnik replied:
> Did you test this logic with the Surrogate characters (well and malformed characters)?

I was relying on the existing tests and the edge cases that I've found and added in https://github.com/dotnet/corefx/pull/41016 and https://github.com/dotnet/corefx/pull/41017

I am going to send one more PR with the Surrogate characters and malformed ones.

tarekgh commented:
> I don't think the ordinal optimization is valid when performing a culture-aware comparison. Digraphs and denormalized forms of accented characters are two places where this could show up.

The optimization here is not ordinal. It is just enumerating over the strings as text items which I believe should be OK as long as it is not applied on cases like ignore non spaces which can break the logic. So, the optimization is not applied on such cases anyway. 

The only thing is, do we think it is worth it adding this complexity to the code for this perf gain? 

@adamsitnik are you planning to remove IsFaseSort too?

adamsitnik commented:
> The only thing is, do we think it is worth it adding this complexity to the code for this perf gain?

I agree that the complexity does not come for free, but it really improves the perf for all non-US cultures when we are searching for a prefix in a big string that does not start with the given prefix.

Example:

```cs
    public class Perf_StartsWith
    {
        public IEnumerable<object[]> StartsWithArguments()
        {
            yield return new object[] { new string('a', 2048), "i" };
        }

        [GlobalSetup]
        public void Setup() => Thread.CurrentThread.CurrentCulture = CultureInfo.GetCultureInfo("fr-FR"); // isAsciiEqualityOrdinal=false

        [Benchmark]
        [ArgumentsSource(nameof(StartsWithArguments))]
        public bool StartsWith(string text, string prefix) => text.StartsWith(prefix);
    }
```

Before: 90,562.510 ns
After:     474.823 ns 

> are you planning to remove IsFaseSort too?

Yes. I also do plan to apply similar optimizations to `EndsWith` once I am done with `StartsWith`.

adamsitnik commented:
@tarekgh thanks for a great code review!

tarekgh commented:
CC @ahsonkhan as he mentioned he had a test case which we can use in the validation too.

tarekgh commented:
https://github.com/dotnet/coreclr/pull/26759
https://github.com/dotnet/coreclr/issues/26758

adamsitnik commented:
@tarekgh  I've added surrogate and malformed Unicode test cases in https://github.com/dotnet/corefx/pull/41227 and all tests are passing. 

Could you PTAL?

tarekgh commented on src/corefx/System.Globalization.Native/pal_collation.c:
```diff
@@ -588,47 +592,164 @@ int32_t GlobalizationNative_IndexOfOrdinalIgnoreCase(
+    switch (strength)
+    {
+        case UCOL_PRIMARY:
+            return UCOL_PRIMARYORDERMASK;
```
do we need to assert on the primary strength? as I mentioned in other comment we'll not expect this strength in the simple case.

adamsitnik replied:
Personally, I am not a big fan of asserts, I prefer to add a lot of unit tests. If we ever set the strength to `UCOL_PRIMARY` all the tests are going to fail so I don't think that we need it.

tarekgh commented on src/corefx/System.Globalization.Native/pal_collation.c:
```diff
@@ -588,47 +592,164 @@ int32_t GlobalizationNative_IndexOfOrdinalIgnoreCase(
+
+        if (patternItem == UCOL_NULLORDER)
+        {
+            if (strength >= UCOL_SECONDARY && sourceItem != UCOL_NULLORDER)
```
do we really need the check strength >= UCOL_SECONDARY? I guess not.

adamsitnik replied:
the check is outside the hot-path and I would prefer to keep it in case for some reason we use it in the future

tarekgh replied:
I am not worried about the perf,  I am worried about the future editing of the code. I'll leave it to you to decide.

tarekgh commented on src/corefx/System.Globalization.Native/pal_collation.c:
```diff
@@ -588,47 +592,164 @@ int32_t GlobalizationNative_IndexOfOrdinalIgnoreCase(
+                if (((sourceItem & UCOL_PRIMARYORDERMASK) == 0) && (sourceItem & UCOL_SECONDARYORDERMASK) != 0)
                 {
-                    result = CanIgnoreAllCollationElements(pColl, lpSource, idx);
+                    return FALSE; // the next non-ignorable character in source text is a combining character
```
could you please try the case when having "o\u0000\u0308"?  thanks.

tarekgh commented:
> Personally, I am not a big fan of asserts, I prefer to add a lot of unit tests. If we ever set the strength to UCOL_PRIMARY all the tests are going to fail so I don't think that we need it.

I am afraid in the future if anyone added any code break this logic will not be easy to catch. test cases may not be enough as we'll not know what case we are not covering can break.

adamsitnik commented:
> I am afraid in the future if anyone added any code break this logic will not be easy to catch. test cases may not be enough as we'll not know what case we are not covering can break.

I think that we have really good test coverage now. Also, the Asserts are executed only for Checked builds? I always build everything in Release when I work on perf so they are never executed for me.

If you really want me to add an assert I can add it for the method that creates the Collators to make sure that we never end up with one that has strength=UCOL_PRIMARY

tarekgh commented:
> If you really want me to add an assert I can add it for the method that creates the Collators to make sure that we never end up with one that has strength=UCOL_PRIMARY

I'll leave it to you. what I am trying to say, if we are going with the simple path we shouldn't having UCOL_PRIMARY strength. This is more asserting non UCOL_PRIMARY strength with simple path. It is not really about how we create the collators. Thanks @adamsitnik for going through all of this.

eerhardt commented:
Just my 2 cents on asserts.

Sometimes the value of the assert is for new people coming into the code and reading it for the first time. It is a nice way to document "this should always hold true". If at a future time it isn't true, the new person knows that the original author never anticipated this situation.

ahsonkhan commented:
> CC @ahsonkhan as he mentioned he had a test case which we can use in the validation too.

I'll dig up the scenario and share.

adamsitnik commented:
@tarekgh I've added proper handling for things like ` "o\u0000\u0308".StartsWith("o")` (great catch BTW!). Everything works as expected, including the test case sent offline my Ahson.

I've also added the asserts.

tarekgh commented:
@adamsitnik  thanks for all tests you are doing. this increase our confidence!

am11 commented on src/corefx/System.Globalization.Native/pal_collation.c:
```diff
@@ -587,6 +591,142 @@ int32_t GlobalizationNative_IndexOfOrdinalIgnoreCase(
+    }
+}
+
+static int32_t SimpleStartsWith_Iterators(UCollationElements* pPatternIterator, UCollationElements* pSourceIterator, UColAttributeValue strength)
```
Could these functions return 1-byte `bool` type defined in C99 `stdbool.h` (already included on top of this file)?

eerhardt replied:
See https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/interop-guidelines.md#unix-shims

> Stick to data types which are guaranteed not to vary in size across flavors.

All of our existing shims are using `int32_t` for booleans, which guarantees the size will be consistent.

adamsitnik commented:
@tarekgh all the tests are passing, do you think that the PR is ready to merge? I would like to apply similar optimization to EndsWith in a separate PR.

tarekgh commented:
Thanks @adamsitnik

adamsitnik commented:
@tarekgh thanks for all the great reviews, test cases and patience to my Unicode learning process!
